Add test template example for helm create

hickeyma opened the pull request:
Test template example for `helm create.` 

Closes #3559

hickeyma committed:
Add test template example for helm create

As helm create ganerates an nginx chart, the example is based on
https://github.com/helm/helm/blob/master/docs/examples/nginx/templates/
service-test.yaml

Signed-off-by: Martin Hickey <martin.hickey@ie.ibm.com>

thomastaylor312 commented:
cc @michelleN as you may be interested in adding a test like this for `helm create`

henrynash commented on cmd/helm/create_test.go:
```diff
@@ -143,8 +143,8 @@ func TestCreateStarterCmd(t *testing.T) {
 
-	if l := len(c.Templates); l != 6 {
-		t.Errorf("Expected 5 templates, got %d", l)
+	if l := len(c.Templates); l != 7 {
```
Although I agree that this change keeps the spirit of the original code, might it be better to define the number of expected templates as a constant and reference to the constant+1 in the test?

bacongobbler replied:
Generally speaking if it's not a core part of the PR (in this case, if it does not add a test template to `helm create`), we ask contributors to file a separate PR. It's a good suggestion, but I wouldn't say it's something that needs to be done to accomplish what this PR is trying to solve.

However, if you feel like submitting a new PR to improve the tests, that would be much appreciated @henrynash.

henrynash replied:
Agree entirely with this approach. I'll submit a follow-on PR.

hickeyma replied:
Ok, thanks for feedback @bacongobbler. @henrynash going to leave it as is, as its not related to the PR and not an issue either.

henrynash reviewed (commented):
minor suggestion regarding number of templates referenced in test

hickeyma commented:
/cc @michelleN Would appreciate your feedback, thanks.

unguiculus commented on pkg/chartutil/create.go:
```diff
@@ -290,6 +292,26 @@ Create chart name and version as used by the chart label.
+    heritage: {{ .Release.Service }}
+    release: {{ .Release.Name }}
+    chart: {{ .Chart.Name }}-{{ .Chart.Version }}
+    app: {{ template "<CHARTNAME>.name" . }}
```
Labels should match those of other resources and `include` should be preferred over `template`:

```yaml
app.kubernetes.io/name: {{ include "<CHARTNAME>.name" . }}
helm.sh/chart: {{ include "<CHARTNAME>.chart" . }}
app.kubernetes.io/instance: {{ .Release.Name }}
app.kubernetes.io/managed-by: {{ .Release.Service }}
```

hickeyma replied:
Good catch. Updated.

hickeyma committed:
Update after review

- https://github.com/helm/helm/pull/4524#pullrequestreview-155183390
- https://github.com/helm/helm/pull/4524#pullrequestreview-155379922
- Fix where app.kubernetes.io/instance was incorrect in service selector

Signed-off-by: Martin Hickey <martin.hickey@ie.ibm.com>

hickeyma commented on pkg/chartutil/create.go:
```diff
@@ -44,8 +44,10 @@ const (
+	// HelpersName is the name of the example helpers file.
 	HelpersName = "_helpers.tpl"
+	// ServiceTestName is the name of the example test service file.
+	ServiceTestName = "service-test.yaml"
```
Sure, done.

hickeyma commented on pkg/chartutil/create.go:
```diff
@@ -290,6 +292,26 @@ Create chart name and version as used by the chart label.
+spec:
+  containers:
+    - name: curl
+      image: radial/busyboxplus:curl
```
Agree, changed to busybox `wget`.

hickeyma commented on pkg/chartutil/create.go:
```diff
@@ -399,6 +421,11 @@ func Create(chartfile *chart.Metadata, dir string) (string, error) {
 			content: Transform(defaultHelpers, "<CHARTNAME>", chartfile.Name),
 		},
+		{
+			// service-test.yaml
```
Done.

hickeyma commented on pkg/chartutil/create.go:
```diff
@@ -399,6 +421,11 @@ func Create(chartfile *chart.Metadata, dir string) (string, error) {
 		},
+		{
+			// service-test.yaml
+			path:    filepath.Join(cdir, TemplatesDir, ServiceTestName),
```
Yep, done.

hickeyma commented:
Thanks @henrynash @bacongobbler @unguiculus for review comments.

Updates done and ready for review again.

bacongobbler commented:
Mind taking a look at the failing tests? Try rebasing first and see if that fixes a few of the tests.

hickeyma committed:
Add doc update

Signed-off-by: Martin Hickey <martin.hickey@ie.ibm.com>

hickeyma committed:
Update GO formatting

Signed-off-by: Martin Hickey <martin.hickey@ie.ibm.com>

bacongobbler commented:
thanks. I'll give this a spin later this week. Thanks!

hickeyma commented:
@bacongobbler @unguiculus Ready for review again when you get a chance.

bacongobbler reviewed (approved):
works as tested.

```
$ helm create foo
$ helm install ./foo
$ helm test $(helm last)
RUNNING: veering-ladybug-foo-test-connection
PASSED: veering-ladybug-foo-test-connection
```

hickeyma commented:
Thanks @bacongobbler

hickeyma commented:
ported in #6490
